### PR TITLE
Retrieve TLS config from apiserver for OpenShift

### PR DESCRIPTION
**Pull request by dkwon17**
### What does this PR do?
* Retrieve TLS config from api server for Openshift (no changes for Kubernetes).
  * Example PR: https://github.com/openshift/cluster-machine-approver/pull/286/changes#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261.
* `openshift/api` dependency bump was needed to use the `configv1.TLSAdherencePolicy` field 

### Is it tested? How?

To test this PR, I used this catalog source to install DWO on an OCP 5.0 cluster:
```
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: dwo-tls-adherence
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: quay.io/dkwon17/devworkspace-operator-index:tls-adherence
  displayName: "DWO TLS Adherence"
  publisher: "Test"
  updateStrategy:
    registryPoll:
      interval: 15m
```
For verification, I followed these testing steps: https://gist.github.com/dkwon17/92211bd5cf6a7a101d1e27c3179acb7a

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

## Summary by CodeRabbit

* **New Features**
  * Added OpenShift-aware TLS configuration for metrics and webhook servers.
  * Servers now gracefully restart when relevant TLS security profiles or policies change.
  * Applied cluster TLS profiles while preserving safe defaults on other platforms.

* **Bug Fixes**
  * Updated permissions to support cluster-wide API server discovery while restricting detailed access appropriately.
  * Improved handling of TLS configuration errors during startup.

* **Tests**
  * Added coverage for TLS profiles, adherence policies, platform differences, restart behavior, and configuration edge cases.

### CODE REVIEW

File: TLS_ADHERENCE_TEST_PLAN.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
+# TLS Adherence Feature Test Plan
```
Comment (rohanKanojia) [inline TLS_ADHERENCE_TEST_PLAN.md]: Shall we use test plan in https://gist.github.com/dkwon17/92211bd5cf6a7a101d1e27c3179acb7a or this one?

File: pkg/tlssetup/server_tls.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+// RegisterSecurityProfileWatcher watches the APIServer TLS profile and adherence policy.
+// Calls onCancel to trigger restart when either changes. No-op on non-OpenShift.
+func RegisterSecurityProfileWatcher(mgr manager.Manager, serverTLS ServerTLS, onCancel context.CancelFunc, log logr.Logger) error {
+	if !infrastructure.IsOpenShift() {
+		return nil
+	}
+
+	// Only set up the watcher if we successfully fetched the initial profile
+	if len(serverTLS.TLSOpts) == 0 {
```
Comment (rohanKanojia) [inline pkg/tlssetup/server_tls.go]: `TLSOpts` ends up empty in two cases: when a fetch failed, and when the fetch succeeded but the policy was "" / `LegacyAdheringComponentsOnly`. Skipping the watcher makes sense in the first case since there's no valid baseline to diff against. In the second, though, I think we'd still want to watch, so we'd catch an admin later setting `StrictAllComponents` ; otherwise that change wouldn't take effect until the pod happens to restart.

Is that scenario already handled somewhere upstream that I'm missing?

File: controllers/workspace/devworkspace_controller.go [hunk 3]
```diff
@@ -96,6 +96,7 @@ type DevWorkspaceReconciler struct {
 // +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=clusterroles;clusterrolebindings,verbs=get;list;watch;create;update
 // +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=roles;rolebindings,verbs=get;list;watch;create;update;delete
 // +kubebuilder:rbac:groups=oauth.openshift.io,resources=oauthclients,verbs=get;list;watch;create;update;patch;delete;deletecollection
+// +kubebuilder:rbac:groups=config.openshift.io,resources=apiservers,verbs=get;list;watch
```
Comment (rohanKanojia) [inline controllers/workspace/devworkspace_controller.go]: Since only the singleton cluster APIServer is ever read, consider scoping with `resourceNames=cluster` (similar to the existing proxies marker on line 101) to follow least-privilege

File: pkg/tlssetup/server_tls.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		return false
+	default:
+		// StrictAllComponents or unknown future value → honor the profile
+		return true
+	}
+}
+
+// BuildServerTLSOptions fetches TLS settings from the OpenShift API server.
+// Only applies the cluster profile when the tlsAdherence policy requires it.
+func BuildServerTLSOptions(ctx context.Context, cfg *rest.Config, scheme *k8sruntime.Scheme, log logr.Logger) (ServerTLS, error) {
```
Comment (rohanKanojia) [inline pkg/tlssetup/server_tls.go]: `BuildServerTLSOptions()` currently logs errors and falls back to Go defaults, but always returns `nil` error. Could we either propagate these errors or remove the error return value?

File: webhook/main.go [hunk 5]
```diff
@@ -89,6 +88,13 @@ func main() {
 		os.Exit(1)
 	}
 
+	serverTLS, err := tlssetup.BuildServerTLSOptions(
```
Comment (rohanKanojia) [inline webhook/main.go:94]: Could we ensure `configv1.AddToScheme(scheme)` is registered in the webhook entrypoint as well?

File: pkg/tlssetup/server_tls_test.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	defer infrastructure.InitializeForTesting(infrastructure.Kubernetes)
+
+	log := zap.New(zap.UseDevMode(true))
+
+	// When TLSOpts is empty (profile not applied), should skip watcher setup and return nil
+	err := RegisterSecurityProfileWatcher(nil, ServerTLS{}, nil, log)
+	if err != nil {
+		t.Errorf("RegisterSecurityProfileWatcher() with empty TLSOpts should skip setup, got error = %v", err)
+	}
+}
```
Comment (rohanKanojia) [inline pkg/tlssetup/server_tls_test.go:91]: Could we also add tests for `BuildServerTLSOptions` covering the main OpenShift paths?

File: controllers/workspace/devworkspace_controller.go (hunk 3 above)
Comment (rohanKanojia) [reply controllers/workspace/devworkspace_controller.go]: the OLM bundle CSV (`deploy/bundle/manifests/devworkspace-operator.clusterserviceversion.yaml`) wasn't regenerated, so it still grants apiservers only verbs: [`get`] (grouped with proxies).

Could you please run `make generate_all` again?

File: TLS_ADHERENCE_TEST_PLAN.md (hunk 1 above)
Comment (dkwon17) [reply TLS_ADHERENCE_TEST_PLAN.md]: Oops, I did not mean to commit this, I will remove it. I would follow what's mentioned in https://gist.github.com/dkwon17/92211bd5cf6a7a101d1e27c3179acb7a

File: pkg/tlssetup/server_tls.go (hunk 2 above)
Comment (rohanKanojia) [reply pkg/tlssetup/server_tls.go]: Is manual oc delete pod in Tests 2 and 5 intended? Or is this a workaround for `RegisterSecurityProfileWatcher` skipping setup when TLSOpts is empty? Asking because it means the watcher can detect leaving strict adherence but not entering it, so an admin turning this on doesn't get it until someone restarts the operator.

File: controllers/workspace/devworkspace_controller.go (hunk 3 above)
Comment (dkwon17) [reply controllers/workspace/devworkspace_controller.go]: Good catch, fixed both issues in my latest commits.

File: pkg/tlssetup/server_tls.go (hunk 4 above)
Comment (dkwon17) [reply pkg/tlssetup/server_tls.go]: I opted to remove the return value and fallback on go defaults.

File: pkg/tlssetup/server_tls.go (hunk 2 above)
Comment (dkwon17) [reply pkg/tlssetup/server_tls.go]: With the latest changes, the watcher is not skipped anymore when the policy is an empty string or `LegacyAdheringComponentsOnly`.

The watcher is still skipped if the initial profile fetch fails however, which acts as a fallback to how DWO functioned before this PR

File: pkg/tlssetup/server_tls_test.go (hunk 6 above)
Comment (dkwon17) [reply pkg/tlssetup/server_tls_test.go:91]: Sounds good, updated

File: webhook/main.go (hunk 5 above)
Comment (dkwon17) [reply webhook/main.go:94]: Updated in my new commits

File: pkg/tlssetup/server_tls_test.go [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+
+	log := zap.New(zap.UseDevMode(true))
+
+	err := RegisterSecurityProfileWatcher(nil, ServerTLS{}, nil, log)
+	if err != nil {
+		t.Errorf("RegisterSecurityProfileWatcher() on Kubernetes should be no-op, got error = %v", err)
+	}
+}
+
+func TestRegisterSecurityProfileWatcher_NoTLSOpts(t *testing.T) {
```
Comment (rohanKanojia) [inline pkg/tlssetup/server_tls_test.go]:
```suggestion
func TestRegisterSecurityProfileWatcher_ProfileNotFetched(t *testing.T) {
```

**Comment by rohanKanojia** [comment]
I tested provided test plan and can confirm it works as expected ✅

**Review by rohanKanojia — APPROVED** [review]

File: pkg/tlssetup/server_tls_test.go (hunk 7 above)
Comment (dkwon17) [reply pkg/tlssetup/server_tls_test.go]: Updated in my latest commit

**Review by btjd — APPROVED** [review]
In webhook/main.go, the old signal handling via os/signal and syscall.SIGTERM is removed and replaced with the cancellable context pattern. The old code created a shutdownChan that was never actually wired to stop the manager, so this looks like a bug fix on top of the TLS work. Do you think that this should be called out explicitly in the PR description?
Other than that it looks good to me.
